File: kaji/ts/src/runtime/cancellation.ts
import type { Clock, TimerHandle, TimerScheduler } from "@/internal/uuid";
import {
  attachProviderViolationSettlement,
  ProviderCancellationContractViolation,
  TurnTimeoutError,
} from "@/runtime/limits";

/**
 * Cancellation token for the agent loop, mirroring the Python
 * `CancellationToken`. Carries a backing `AbortController` so the same
 * cancel signal can be plumbed into platform APIs that accept an
 * `AbortSignal` (the `openai` / `@anthropic-ai/sdk` clients, `fetch`,
 * `EventTarget` listeners, etc.). The boolean `isCancelled` flag remains
 * for the polling style the runtime already uses; both fire together.
 */
export class CancellationError extends Error {
  constructor(message = "Agent run was cancelled") {
    super(message);
    this.name = "CancellationError";
  }
}

export interface CancellationTokenLike {
  isCancelled: boolean;
  signal?: AbortSignal;
  throwIfCancelled?: () => void;
}

export function throwIfCancellationRequested(token?: CancellationTokenLike): void {
  if (!token?.isCancelled) return;
  if (typeof token.throwIfCancelled === "function") {
    token.throwIfCancelled();
  }
  throw new CancellationError();
}

function isCancellationCompatible(error: unknown): boolean {
  return (
    error instanceof CancellationError ||
    (typeof error === "object" && error !== null && "name" in error && error.name === "AbortError")
  );
}

export class CancellationToken {
  private readonly controller = new AbortController();

  /** Whether `cancel()` has been called. */
  get isCancelled(): boolean {
    return this.controller.signal.aborted;
  }

  /**
   * The underlying `AbortSignal`. Pass this directly to APIs that accept
   * one (the OpenAI / Anthropic SDKs, `fetch`, etc.) so the network call
   * aborts on cancel instead of just polling out at the next yield point.
   */
  get signal(): AbortSignal {
    return this.controller.signal;
  }

  cancel(): void {
    if (!this.controller.signal.aborted) {
      this.controller.abort();
    }
  }

  throwIfCancelled(): void {
    if (this.controller.signal.aborted) {
      throw new CancellationError();
    }
  }
}

/** Disposable linked deadline scope that owns one provider iterator. */
export class DeadlineCancellationScope {
  readonly token = new CancellationToken();
  private readonly deadlineWake: Promise<void>;
  private readonly parentWake: Promise<void>;
  private resolveDeadline!: () => void;
  private resolveParent!: () => void;
  private readonly parentAbort: () => void;
  private readonly deadlineTimer: TimerHandle;
  private iterator: AsyncIterator<unknown> | undefined;
  private activeNext: Promise<IteratorResult<unknown>> | undefined;
  private closePromise: Promise<void> | undefined;
  private transferred = false;
  private yielded = false;
  private dispatched = false;
  private cancellationRequestedAtMs: number | undefined;
  private cancellationSource: "parent" | "deadline" | "provider" | undefined;

  constructor(
    private readonly parent: CancellationToken,
    private readonly deadlineMonotonicMs: number,
    private readonly cancellationGraceMs: number,
    private readonly clock: Clock,
    private readonly scheduler: TimerScheduler,
  ) {
    if (!Number.isFinite(deadlineMonotonicMs) || deadlineMonotonicMs < 0) {
      throw new RangeError("deadlineMonotonicMs must be a finite non-negative number");
    }
    if (!Number.isFinite(cancellationGraceMs) || cancellationGraceMs <= 0) {
      throw new RangeError("cancellationGraceMs must be a positive finite number");
    }
    this.deadlineWake = new Promise((resolve) => {
      this.resolveDeadline = resolve;
    });
    this.parentWake = new Promise((resolve) => {
      this.resolveParent = resolve;
    });
    this.parentAbort = () => {
      this.requestCancellation("parent");
      this.resolveParent();
    };
    parent.signal.addEventListener("abort", this.parentAbort, { once: true });
    if (parent.isCancelled) this.parentAbort();
    this.deadlineTimer = scheduler.schedule(
      Math.max(0, deadlineMonotonicMs - clock.nowMonotonic()),
      () => {
        this.requestCancellation("deadline");
        this.resolveDeadline();
      },
    );
  }

  async *consume<T>(stream: AsyncIterable<T>): AsyncGenerator<T> {
    this.iterator = stream[Symbol.asyncIterator]() as AsyncIterator<unknown>;
    let providerError: unknown;
    try {
      while (true) {
        if (this.parent.isCancelled) await this.abort<T>(undefined, true, "parent");
        if (this.clock.nowMonotonic() >= this.deadlineMonotonicMs) {
          await this.abort<T>(undefined, false, "deadline");
        }
        if (this.token.isCancelled && this.cancellationSource === undefined) {
          await this.abort<T>(undefined, true, "provider");
        }

        this.dispatched = true;
        const next = this.iterator.next() as Promise<IteratorResult<T>>;
        this.activeNext = next as Promise<IteratorResult<unknown>>;
        const settlementEvidence = () => ({
          settledAtMs: this.clock.nowMonotonic(),
          cancellationSource: this.cancellationSource,
          parentCancelled: this.parent.isCancelled,
          tokenCancelled: this.token.isCancelled,
        });
        const providerResult = next.then(
          (result) => ({ status: "settled" as const, result, ...settlementEvidence() }),
          (error: unknown) => ({ status: "failed" as const, error, ...settlementEvidence() }),
        );
        const wake = await Promise.race([
          providerResult,
          this.deadlineWake.then(() => ({ status: "deadline" as const })),
          this.parentWake.then(() => ({ status: "parent" as const })),
        ]);

        if (wake.status === "failed") {
          if (
            (wake.parentCancelled || wake.cancellationSource === "parent") &&
            isCancellationCompatible(wake.error)
          ) {
            await this.abort<T>(next, true, "parent");
          }
          if (
            wake.cancellationSource === "deadline" ||
            wake.settledAtMs >= this.deadlineMonotonicMs
          ) {
            await this.abort<T>(next, false, "deadline");
          }
          if (
            wake.cancellationSource === "provider" ||
            (wake.tokenCancelled && wake.cancellationSource === undefined)
          ) {
            await this.abort<T>(next, true, "provider");
          }
          this.activeNext = undefined;
          throw wake.error;
        }
        if (this.parent.isCancelled) await this.abort<T>(next, true, "parent");
        if (wake.status === "deadline" || this.clock.nowMonotonic() >= this.deadlineMonotonicMs) {
          await this.abort<T>(next, false, "deadline");
        }
        if (this.token.isCancelled && this.cancellationSource === undefined) {
          await this.abort<T>(next, true, "provider");
        }
        if (wake.status !== "settled") continue;
        this.activeNext = undefined;
        if (wake.result.done) return;
        this.yielded = true;
        yield wake.result.value;
      }
    } catch (error) {
      if (error instanceof ProviderCancellationContractViolation) {
        this.transferred = true;
      } else {
        providerError = error;
      }
      throw error;
    } finally {
      if (!this.transferred) {
        try {
          await this.finishClose();
        } catch (closeError) {
          if (closeError instanceof ProviderCancellationContractViolation) {
            this.transferred = true;
            if (providerError !== undefined && closeError.cause === undefined) {
              Object.defineProperty(closeError, "cause", {
                value: providerError,
                configurable: true,
              });
            }
            throw closeError;
          }
          if (providerError === undefined) throw closeError;
        }
      }
    }
  }

  dispose(): void {
    this.deadlineTimer.cancel();
    this.parent.signal.removeEventListener("abort", this.parentAbort);
  }

  private async abort<T>(
    active: Promise<IteratorResult<T>> | undefined,
    callerCancelled: boolean,
    source: "parent" | "deadline" | "provider",
  ): Promise<never> {
    this.requestCancellation(source);
    const settlement = Promise.resolve(active)
      .catch(() => undefined)
      .then(() => this.closeOnce());
    let graceTimer: TimerHandle | undefined;
    const grace = new Promise<{ status: "grace" }>((resolve) => {
      const requestedAt = this.cancellationRequestedAtMs ?? this.clock.nowMonotonic();
      const remaining = Math.max(
        0,
        requestedAt + this.cancellationGraceMs - this.clock.nowMonotonic(),
      );
      graceTimer = this.scheduler.schedule(remaining, () => resolve({ status: "grace" }));
    });
    try {
      const result = await Promise.race([
        settlement.then(
          () => ({ status: "settled" as const }),
          () => ({ status: "failed" as const }),
        ),
        grace,
      ]);
      if (result.status === "settled") {
        if (callerCancelled) throw new CancellationError();
        throw new TurnTimeoutError(
          this.yielded ? "provider_stream" : "provider_open",
          true,
          this.dispatched ? "unknown" : "not_started",
        );
      }
      if (result.status === "failed") {
        this.transferred = true;
        throw this.contractViolation(settlement);
      }
      this.transferred = true;
      throw this.contractViolation(settlement);
    } finally {
      graceTimer?.cancel();
    }
  }

  private closeOnce(): Promise<void> {
    if (this.closePromise !== undefined) return this.closePromise;
    const iterator = this.iterator;
    this.closePromise = (async () => {
      if (this.activeNext !== undefined) {
        try {
          await this.activeNext;
        } catch {
          // Provider failure is surfaced by the active next owner.
        }
      }
      await iterator?.return?.();
    })();
    return this.closePromise;
  }

  private async finishClose(): Promise<void> {
    if (this.iterator === undefined) return;
    const close = this.closeOnce();
    const settled = close.then(
      () => ({ status: "settled" as const }),
      () => ({ status: "failed" as const }),
    );
    const wake = await Promise.race([
      settled,
      this.deadlineWake.then(() => ({ status: "deadline" as const })),
      this.parentWake.then(() => ({ status: "parent" as const })),
    ]);
    if (wake.status === "settled") return;
    if (wake.status === "failed") {
      this.transferred = true;
      throw this.contractViolation(close);
    }
    if (this.parent.isCancelled) await this.abort(this.activeNext, true, "parent");
    if (wake.status === "deadline" || this.clock.nowMonotonic() >= this.deadlineMonotonicMs) {
      await this.abort(this.activeNext, false, "deadline");
    }
    await close;
  }

  private requestCancellation(source: "parent" | "deadline" | "provider"): void {
    if (this.cancellationRequestedAtMs === undefined) {
      this.cancellationRequestedAtMs = this.clock.nowMonotonic();
    }
    this.cancellationSource ??= source;
    this.token.cancel();
  }

  private contractViolation(settlement: Promise<void>): ProviderCancellationContractViolation {
    const phase = this.yielded ? "provider_stream" : "provider_open";
    return attachProviderViolationSettlement(
      new ProviderCancellationContractViolation(phase),
      settlement,
    );
  }
}

export function createDeadlineCancellationScope(
  parent: CancellationToken,
  deadlineMonotonicMs: number,
  cancellationGraceMs: number,
  clock: Clock,
  scheduler: TimerScheduler,
): DeadlineCancellationScope {
  return new DeadlineCancellationScope(
    parent,
    deadlineMonotonicMs,
    cancellationGraceMs,
    clock,
    scheduler,
  );
}

File: kaji/packages/py/src/runtime/agents/cancellation.py
import asyncio
from collections.abc import AsyncIterator
from contextlib import suppress
import math
from typing import Any, Generic, TypeVar

from kaji.runtime.agents.limits import (
    ProviderCancellationContractViolation,
    TurnTimeoutError,
)
from kaji.core.determinism import Clock, TimerScheduler


T = TypeVar("T")


class CancelledError(asyncio.CancelledError):
    """Raised by :meth:`CancellationToken.raise_if_cancelled` when the
    token has been triggered. Subclasses :class:`asyncio.CancelledError`
    so callers can ``except asyncio.CancelledError`` uniformly across the
    SDK and asyncio's own cancellation machinery."""


class CancellationToken:
    """Token to coordinate cancellation across asynchronous boundaries.

    Wraps an :class:`asyncio.Event` so the token composes with asyncio's
    primitives. Use ``token.is_cancelled`` for cheap polling, ``await
    token.wait()`` to integrate with ``asyncio.wait_for`` / select-style
    races, and ``token.raise_if_cancelled()`` at yield points to fail
    fast with :class:`asyncio.CancelledError`.
    """

    def __init__(self) -> None:
        self._event = asyncio.Event()

    def cancel(self) -> None:
        """Trigger the cancellation. Idempotent."""
        self._event.set()

    @property
    def is_cancelled(self) -> bool:
        """Check if cancellation has been requested."""
        return self._event.is_set()

    @property
    def event(self) -> asyncio.Event:
        """The underlying :class:`asyncio.Event`. Useful for ``await
        token.event.wait()`` or composing with ``asyncio.wait`` to race
        the token against other futures."""
        return self._event

    async def wait(self) -> None:
        """Block until ``cancel()`` is called. Returns immediately if the
        token is already cancelled."""
        await self._event.wait()

    def raise_if_cancelled(self) -> None:
        """Raise :class:`asyncio.CancelledError` if the token is set.
        Use this at yield points (between I/O steps) to participate in
        asyncio's structured cancellation."""
        if self._event.is_set():
            raise CancelledError("Agent run was cancelled")


class ProviderDeadlineScope(Generic[T]):
    """Own one provider iterator, deadline timer, and cooperative shutdown."""

    def __init__(
        self,
        *,
        parent: CancellationToken,
        deadline_monotonic: float,
        cancellation_grace_seconds: float,
        clock: Clock,
        scheduler: TimerScheduler,
    ) -> None:
        if isinstance(deadline_monotonic, bool) or not isinstance(
            deadline_monotonic, (int, float)
        ):
            raise TypeError("deadline_monotonic must be a finite non-negative number")
        if not math.isfinite(float(deadline_monotonic)) or deadline_monotonic < 0:
            raise ValueError("deadline_monotonic must be a finite non-negative number")
        if isinstance(cancellation_grace_seconds, bool) or not isinstance(
            cancellation_grace_seconds, (int, float)
        ):
            raise TypeError("cancellation_grace_seconds must be a positive number")
        if (
            not math.isfinite(float(cancellation_grace_seconds))
            or cancellation_grace_seconds <= 0
        ):
            raise ValueError("cancellation_grace_seconds must be a positive number")
        self.parent = parent
        self.token = CancellationToken()
        self.deadline_monotonic = deadline_monotonic
        self.cancellation_grace_seconds = cancellation_grace_seconds
        self.clock = clock
        self.scheduler = scheduler
        self._parent_wait: asyncio.Task[None] | None = None
        self._deadline_wait: asyncio.Future[None] | None = None
        self._deadline_timer: Any = None
        self._iterator: AsyncIterator[T] | None = None
        self._active_next: asyncio.Task[T] | None = None
        self._close_task: asyncio.Task[None] | None = None
        self._transferred = False
        self._yielded = False
        self._dispatched = False
        self._cancellation_requested_at: float | None = None
        self._cancellation_source: str | None = None

    async def __aenter__(self) -> "ProviderDeadlineScope[T]":
        loop = asyncio.get_running_loop()
        self._parent_wait = asyncio.create_task(self.parent.wait())
        self._deadline_wait = loop.create_future()
        self._deadline_timer = self.scheduler.call_later(
            max(0.0, self.deadline_monotonic - self.clock.now_monotonic()),
            self._request_deadline_cancellation,
        )
        return self

    async def __aexit__(self, *exc_info: object) -> None:
        if not self._transferred:
            await self._finish_close()
        if self._deadline_timer is not None:
            self._deadline_timer.cancel()
        if self._parent_wait is not None:
            self._parent_wait.cancel()
            with suppress(asyncio.CancelledError):
                await self._parent_wait
        if self._deadline_wait is not None and not self._deadline_wait.done():
            self._deadline_wait.cancel()

    async def consume(self, stream: Any) -> AsyncIterator[T]:
        iterator = stream.__aiter__()
        self._iterator = iterator
        provider_error: BaseException | None = None
        try:
            while True:
                if self.parent.is_cancelled:
                    await self._abort(caller_cancelled=True, source="parent")
                if self.clock.now_monotonic() >= self.deadline_monotonic:
                    await self._abort(caller_cancelled=False, source="deadline")
                if self.token.is_cancelled and self._cancellation_source is None:
                    await self._abort(caller_cancelled=True, source="provider")

                settled_at: float | None = None
                settled_source: str | None = None
                settled_parent_cancelled = False
                settled_token_cancelled = False

                async def next_with_evidence() -> T:
                    nonlocal settled_at
                    nonlocal settled_source
                    nonlocal settled_parent_cancelled
                    nonlocal settled_token_cancelled
                    try:
                        return await iterator.__anext__()
                    finally:
                        settled_at = self.clock.now_monotonic()
                        settled_source = self._cancellation_source
                        settled_parent_cancelled = self.parent.is_cancelled
                        settled_token_cancelled = self.token.is_cancelled

                self._dispatched = True
                next_task = asyncio.create_task(next_with_evidence())
                self._active_next = next_task
                assert self._parent_wait is not None
                assert self._deadline_wait is not None
                await asyncio.wait(
                    {next_task, self._parent_wait, self._deadline_wait},
                    return_when=asyncio.FIRST_COMPLETED,
                )

                if next_task.done() and not next_task.cancelled():
                    provider_exception = next_task.exception()
                    if provider_exception is not None and not isinstance(
                        provider_exception, StopAsyncIteration
                    ):
                        assert settled_at is not None
                        if (
                            settled_parent_cancelled or settled_source == "parent"
                        ) and (
                            isinstance(provider_exception, asyncio.CancelledError)
                            or type(provider_exception).__name__ == "AbortError"
                        ):
                            await self._abort(caller_cancelled=True, source="parent")
                        if (
                            settled_source == "deadline"
                            or settled_at >= self.deadline_monotonic
                        ):
                            await self._abort(
                                caller_cancelled=False,
                                source="deadline",
                            )
                        if settled_source == "provider" or (
                            settled_token_cancelled and settled_source is None
                        ):
                            await self._abort(caller_cancelled=True, source="provider")
                        self._active_next = None
                        raise provider_exception

                # Explicit priority for control signals: caller, then deadline.
                if self.parent.is_cancelled:
                    await self._abort(caller_cancelled=True, source="parent")
                if self._deadline_wait.done() or (
                    self.clock.now_monotonic() >= self.deadline_monotonic
                ):
                    await self._abort(caller_cancelled=False, source="deadline")
                if self.token.is_cancelled and self._cancellation_source is None:
                    await self._abort(caller_cancelled=True, source="provider")

                try:
                    chunk = next_task.result()
                except StopAsyncIteration:
                    self._active_next = None
                    return
                self._active_next = None
                self._yielded = True
                yield chunk
        except ProviderCancellationContractViolation:
            self._transferred = True
            raise
        except asyncio.CancelledError:
            current = asyncio.current_task()
            if current is not None and current.cancelling() > 0:
                await self._abort(caller_cancelled=True, source="external")
            raise
        except BaseException as error:
            provider_error = error
            raise
        finally:
            if not self._transferred:
                try:
                    await self._finish_close()
                except asyncio.CancelledError:
                    current = asyncio.current_task()
                    if current is not None and current.cancelling() > 0:
                        await self._abort(caller_cancelled=True, source="external")
                        raise
                    if provider_error is None:
                        raise
                except ProviderCancellationContractViolation as close_error:
                    if provider_error is not None:
                        raise close_error from provider_error
                    raise
                except BaseException:
                    if provider_error is None:
                        raise

    async def _abort(self, *, caller_cancelled: bool, source: str) -> None:
        self._request_cancellation(source)
        active = self._active_next
        if active is not None and not active.done():
            active.cancel()

        cleanup = asyncio.create_task(self._settle_and_close(active))
        grace_wait = asyncio.get_running_loop().create_future()
        requested_at = self._cancellation_requested_at
        assert requested_at is not None
        timer = self.scheduler.call_later(
            max(
                0.0,
                requested_at
                + self.cancellation_grace_seconds
                - self.clock.now_monotonic(),
            ),
            lambda: grace_wait.set_result(None) if not grace_wait.done() else None,
        )
        try:
            await asyncio.wait(
                {cleanup, grace_wait}, return_when=asyncio.FIRST_COMPLETED
            )
            if cleanup.done():
                try:
                    cleanup.result()
                except BaseException:
                    self._transferred = True
                    raise ProviderCancellationContractViolation(
                        settlement=cleanup,
                        phase="provider_stream" if self._yielded else "provider_open",
                    )
                if caller_cancelled:
                    raise CancelledError("Agent run was cancelled")
                raise TurnTimeoutError(
                    phase="provider_stream" if self._yielded else "provider_open",
                    retryable=True,
                    outcome="unknown" if self._dispatched else "not_started",
                )
            self._transferred = True
            raise ProviderCancellationContractViolation(
                settlement=cleanup,
                phase="provider_stream" if self._yielded else "provider_open",
            )
        finally:
            timer.cancel()
            if not grace_wait.done():
                grace_wait.cancel()

    def _request_cancellation(self, source: str) -> None:
        if self._cancellation_requested_at is None:
            self._cancellation_requested_at = self.clock.now_monotonic()
        if self._cancellation_source is None:
            self._cancellation_source = source
        self.token.cancel()

    def _request_deadline_cancellation(self) -> None:
        self._request_cancellation("deadline")
        if self._deadline_wait is not None and not self._deadline_wait.done():
            self._deadline_wait.set_result(None)

    async def _settle_and_close(self, active: asyncio.Task[T] | None) -> None:
        if active is not None:
            with suppress(BaseException):
                await active
            if self._active_next is active:
                self._active_next = None
        await self._close_once()

    async def _close_once(self) -> None:
        if self._iterator is None:
            return
        if self._close_task is None:

            async def close_serialized() -> None:
                active = self._active_next
                if active is not None and not active.done():
                    with suppress(BaseException):
                        await active
                close = getattr(self._iterator, "aclose", None)
                if callable(close):
                    await close()

            self._close_task = asyncio.create_task(close_serialized())
        await self._close_task

    async def _finish_close(self) -> None:
        if self._iterator is None:
            return
        close_task = asyncio.create_task(self._close_once())
        assert self._parent_wait is not None
        assert self._deadline_wait is not None
        await asyncio.wait(
            {close_task, self._parent_wait, self._deadline_wait},
            return_when=asyncio.FIRST_COMPLETED,
        )
        if close_task.done():
            try:
                close_task.result()
            except BaseException:
                self._transferred = True
                raise ProviderCancellationContractViolation(
                    settlement=close_task,
                    phase="provider_stream" if self._yielded else "provider_open",
                )
            return
        if self.parent.is_cancelled:
            await self._abort(caller_cancelled=True, source="parent")
        if self._deadline_wait.done() or (
            self.clock.now_monotonic() >= self.deadline_monotonic
        ):
            await self._abort(caller_cancelled=False, source="deadline")
        await close_task
